File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    bool check(string s, int i, int j) {
        while(i < j) {
            if(s[i++] != s[j--]) return false;
        }
        return true;
    }

    void solve(string s, int idx, vector<string> &curr, vector<vector<string>> &ans) {
        int n = s.size();
        if(idx == n) {
            ans.push_back(curr);
            return;
        }

        for(int i = idx; i<n; i++) {
            if(check(s, idx, i)) {
                curr.push_back(s.substr(idx, i - idx + 1));
                solve(s, i + 1, curr, ans);
                curr.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        if(s.size() == 0) return ans;
        vector<string> curr;
        solve(s, 0, curr, ans);
        return ans;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution(object):
    def check(self, s, i, j):
        while i < j:
            if s[i] != s[j]:
                return False
            i += 1
            j -= 1
        return True

    def solve(self, s, idx, curr, ans):
        n = len(s)
        if idx == n:
            ans.append(list(curr))
            return

        for i in range(idx, n):
            if self.check(s, idx, i):
                curr.append(s[idx:i + 1])
                self.solve(s, i + 1, curr, ans)
                curr.pop()

    def partition(self, s):
        ans = []
        if len(s) == 0:
            return ans
        curr = []
        self.solve(s, 0, curr, ans)
        return ans
        